File: OpenHarmony/developtools/hapsigner/binary_sign_tool/common/src/options.cpp
/*
 * Copyright (c) 2025-2025 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "options.h"

namespace OHOS {
namespace SignatureTools {

/* Initializes the static member constant. */
const std::string Options::KEY_ALIAS = "keyAlias";
const std::string Options::KEY_RIGHTS = "keyPwd";
const std::string Options::KEY_ALG = "keyAlg";
const std::string Options::KEY_SIZE = "keySize";
const std::string Options::KEY_STORE_FILE = "keystoreFile";
const std::string Options::KEY_STORE_RIGHTS = "keystorePwd";
const std::string Options::ISSUER_KEY_ALIAS = "issuerKeyAlias";
const std::string Options::ISSUER_KEY_RIGHTS = "issuerKeyPwd";
const std::string Options::ISSUER_KEY_STORE_FILE = "issuerKeystoreFile";
const std::string Options::ISSUER_KEY_STORE_RIGHTS = "issuerKeystorePwd";
const std::string Options::ISSUER = "issuer";
const std::string Options::SUBJECT = "subject";
const std::string Options::VALIDITY = "validity";
const std::string Options::SIGN_ALG = "signAlg";
const std::string Options::BASIC_CONSTRAINTS_PATH_LEN = "basicConstraintsPathLen";
const std::string Options::OUT_FILE = "outFile";
const std::string Options::OUT_FORM = "outForm";
const std::string Options::SUB_CA_CERT_FILE = "subCaCertFile";
const std::string Options::CA_CERT_FILE = "rootCaCertFile";
const std::string Options::IN_FILE = "inFile";
const std::string Options::PROFILE_CERT_FILE = "profileCertFile";
const std::string Options::APP_CERT_FILE = "appCertFile";
const std::string Options::KEY_USAGE = "keyUsage";
const std::string Options::EXT_KEY_USAGE = "extKeyUsage";
const std::string Options::KEY_USAGE_CRITICAL = "keyUsageCritical";
const std::string Options::EXT_KEY_USAGE_CRITICAL = "extKeyUsageCritical";
const std::string Options::BASIC_CONSTRAINTS_CA = "basicConstraintsCa";
const std::string Options::BASIC_CONSTRAINTS_CRITICAL = "basicConstraintsCritical";
const std::string Options::BASIC_CONSTRAINTS = "basicConstraints";
const std::string Options::OUT_FORM_SCOPE = "cert,certChain";
const std::string Options::MODE = "mode";
const std::string Options::INFORM = "inForm";
const std::string Options::OUT_CERT_CHAIN = "outCertChain";
const std::string Options::OUT_PROFILE = "outProfile";
const std::string Options::PROOF_FILE = "outproof";
const std::string Options::PROFILE_FILE = "profileFile";
const std::string Options::PROFILE_SIGNED = "profileSigned";
const std::string Options::MODULE_FILE = "moduleFile";
const std::string Options::AD_HOC = "adHoc";

char* Options::GetChars(const std::string& key)
{
    if (this->count(key) == 0) {
        return nullptr;
    }

    auto value = (*this)[key];
    char** charsPtr = std::get_if<char*>(&value);
    if (charsPtr == nullptr) {
        SIGNATURE_TOOLS_LOGI("GetChars key = %s  value = nullptr !", key.c_str());
        return nullptr;
    }
    return *charsPtr;
}

std::string Options::GetString(const std::string& key)
{
    if (this->count(key) == 0) {
        return "";
    }

    auto value = (*this)[key];
    std::string* stringPtr = std::get_if<std::string>(&value);
    if (stringPtr == nullptr) {
        SIGNATURE_TOOLS_LOGI("GetString  key = %s value = """, key.c_str());
        return "";
    }
    return *stringPtr;
}

std::string Options::GetString(const std::string& key, const std::string& checkStr)
{
    if (this->count(key) == 0) {
        return "";
    }

    auto value = (*this)[key];
    std::string* stringPtr = std::get_if<std::string>(&value);
    if (stringPtr == nullptr || *stringPtr == "") {
        SIGNATURE_TOOLS_LOGI("GetString  key = %s value = %s ", key.c_str(), checkStr.c_str());
        return checkStr;
    }
    return *stringPtr;
}

int Options::GetInt(const std::string& key)
{
    if (this->count(key) == 0) {
        return 0;
    }

    auto value = (*this)[key];
    int* stringPtr = std::get_if<int>(&value);
    if (stringPtr == nullptr) {
        SIGNATURE_TOOLS_LOGI("GetInt  key = %s value = 0 ", key.c_str());
        return 0;
    }
    return *stringPtr;
}

bool Options::Equals(const std::string& argf, const std::string& args)
{
    std::string ksFile = GetString(argf);
    std::string iksFile = GetString(args);
    if (ksFile == iksFile) {
        return true;
    }
    return  false;
}

bool Options::Required(const std::initializer_list<std::string>& keys)
{
    for (auto& key : keys) {
        if (!this->IsEmpty(key) && !(this->find(key) != this->end())) {
            PrintErrorNumberMsg("COMMAND_ERROR", COMMAND_ERROR, "Params '-" + key + "' is required");
            return false;
        }
    }
    return true;
}

bool Options::IsEmpty(const std::string& cs)
{
    if (cs.empty()) {
        return true;
    }
    return false;
}

bool Options::GetBool(const std::string& key)
{
    auto value = (*this)[key];
    bool* stringPtr = std::get_if<bool>(&value);
    return *stringPtr;
}
} // namespace SignatureTools
} // namespace OHOS

File: OpenHarmony/developtools/hapsigner/hapsigntool/hap_sign_tool_lib/src/main/java/com/ohos/hapsigntool/entity/Options.java
/*
 * Copyright (c) 2021-2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool.entity;

import com.ohos.hapsigntool.error.CustomException;
import com.ohos.hapsigntool.error.ERROR;
import com.ohos.hapsigntool.error.SignToolErrMsg;
import com.ohos.hapsigntool.utils.LogUtils;

import java.util.HashMap;

/**
 * Options parameter class.
 *
 * @since 2021/12/28
 */
public class Options extends HashMap<String, Object> {
    /**
     * App cert file parameter name.
     */
    public static final String APP_CERT_FILE = "appCertFile";

    /**
     * BasicConstraints parameter name.
     */
    public static final String BASIC_CONSTRAINTS = "basicConstraints";

    /**
     * BasicConstraintsCa parameter name.
     */
    public static final String BASIC_CONSTRAINTS_CA = "basicConstraintsCa";

    /**
     * BasicConstraintsCritical parameter name.
     */
    public static final String BASIC_CONSTRAINTS_CRITICAL = "basicConstraintsCritical";

    /**
     * BasicConstraintsPathLen parameter name.
     */
    public static final String BASIC_CONSTRAINTS_PATH_LEN = "basicConstraintsPathLen";

    /**
     * End file type of cert. values in: cert / certChain
     */
    public static final String OUT_FORM = "outForm";

    /**
     * Ca cert file parameter name.
     */
    public static final String CA_CERT_FILE = "rootCaCertFile";

    /**
     * Sub cert for sign
     */
    public static final String SUB_CA_CERT_FILE = "subCaCertFile";

    /**
     * Ext cfg file parameter name.
     */
    public static final String EXT_CFG_FILE = "extCfgFile";

    /**
     * Ext key usage parameter name.
     */
    public static final String EXT_KEY_USAGE = "extKeyUsage";

    /**
     * Ext key usage critical parameter name.
     */
    public static final String EXT_KEY_USAGE_CRITICAL = "extKeyUsageCritical";

    /**
     * In file parameter name.
     */
    public static final String IN_FILE = "inFile";

    /**
     * In file parameter form.
     */
    public static final String IN_FORM = "inForm";

    /**
     * Issuer parameter name.
     */
    public static final String ISSUER = "issuer";

    /**
     * Issuer key alias parameter name.
     */
    public static final String ISSUER_KEY_ALIAS = "issuerKeyAlias";

    /**
     * Issuer key right parameter name.
     */
    public static final String ISSUER_KEY_RIGHTS = "issuerKeyPwd";

    /**
     * Issuer Key Store File parameter name.
     */
    public static final String ISSUER_KEY_STORE_FILE = "issuerKeystoreFile";

    /**
     * Issuer Key Store Pwd parameter name.
     */
    public static final String ISSUER_KEY_STORE_RIGHTS = "issuerKeystorePwd";

    /**
     * Key alg parameter name.
     */
    public static final String KEY_ALG = "keyAlg";

    /**
     * Key alias parameter name.
     */
    public static final String KEY_ALIAS = "keyAlias";

    /**
     * Key right parameter name.
     */
    public static final String KEY_RIGHTS = "keyPwd";

    /**
     * Key size parameter name.
     */
    public static final String KEY_SIZE = "keySize";

    /**
     * Keystore file parameter name.
     */
    public static final String KEY_STORE_FILE = "keystoreFile";

    /**
     * Keystore right parameter name.
     */
    public static final String KEY_STORE_RIGHTS = "keystorePwd";

    /**
     * Key usage parameter name.
     */
    public static final String KEY_USAGE = "keyUsage";

    /**
     * Key usage critical parameter name.
     */
    public static final String KEY_USAGE_CRITICAL = "keyUsageCritical";

    /**
     * Mode parameter name.
     */
    public static final String MODE = "mode";

    /**
     * Out file parameter name.
     */
    public static final String OUT_FILE = "outFile";

    /**
     * Out file parameter name.
     */
    public static final String OUT_CERT_CHAIN = "outCertChain";

    /**
     * Out file parameter name.
     */
    public static final String OUT_PROFILE = "outProfile";

    /**
     * Profile cert file parameter name.
     */
    public static final String PROFILE_CERT_FILE = "profileCertFile";

    /**
     * Profile file parameter name.
     */
    public static final String PROFILE_FILE = "profileFile";

    /**
     * Profile file is signed or not.
     */
    public static final String PROFILE_SIGNED = "profileSigned";

    /**
     * Sign alg parameter name.
     */
    public static final String SIGN_ALG = "signAlg";

    /**
     * Subject parameter name.
     */
    public static final String SUBJECT = "subject";

    /**
     * Trusted app source file parameter name.
     */
    public static final String TRUSTED_APP_SOURCE_FILE = "trustedAppSourceFile";

    /**
     * Trusted root ca file parameter name.
     */
    public static final String TRUSTED_ROOT_CA_FILE = "trustedRootCaFile";

    /**
     * Validity parameter name.
     */
    public static final String VALIDITY = "validity";

    /**
     * All usages included in the extended key usage.
     */
    public static final String EXT_KEY_USAGE_SCOPE = "clientAuthentication,serverAuthentication,codeSignature,"
            + "emailProtection,smartCardLogin,timestamp,ocspSignature";

    /**
     * Key usage includes all usages.
     */
    public static final String KEY_USAGE_SCOPE = "digitalSignature,nonRepudiation,keyEncipherment,dataEncipherment,"
            + "keyAgreement,certificateSignature,crlSignature,encipherOnly,decipherOnly";

    /**
     * Out form includes all forms.
     */
    public static final String OUT_FORM_SCOPE = "cert,certChain";

    /**
     * Serial version UID.
     */
    private static final long serialVersionUID = 1L;

    /**
     * Empty char.
     */
    private static final char[] NO_CHAR = {};

    /**
     * Logger.
     */
    private static final LogUtils LOGGER = new LogUtils(Options.class);


    /**
     * Check required key, throw exception
     *
     * @param keys keys to check
     */
    public void required(String... keys) {
        for (String key : keys) {
            if (!isEmpty(key) && !this.containsKey(key)) {
                CustomException.throwException(ERROR.COMMAND_ERROR, SignToolErrMsg.PARAM_REQUIRED.toString(key));
            }
        }
    }

    private boolean isEmpty(final CharSequence cs) {
        return cs == null || cs.length() == 0;
    }

    /**
     * Get char value of key.
     *
     * @param key key
     * @return value of key
     */
    public char[] getChars(String key) {
        Object value = this.get(key);
        if (value instanceof char[]) {
            return (char[]) value;
        }
        return NO_CHAR;
    }

    /**
     * Get string value of key.
     *
     * @param key      key
     * @param defValue defValue
     * @return string value of key
     */
    public String getString(String key, String defValue) {
        Object value = this.get(key);
        if (!(value instanceof String)) {
            return defValue;
        }
        return (String) value;
    }

    /**
     * Get string value of key or def value.
     *
     * @param key key
     * @return string value of key
     */
    public String getString(String key) {
        return getString(key, "");
    }

    /**
     * Get boolean value of key.
     *
     * @param key      key
     * @param defValue defValue
     * @return boolean value of key
     */
    public boolean getBoolean(String key, boolean defValue) {
        Object value = this.get(key);
        boolean result = defValue;
        if (value instanceof Boolean) {
            result = (boolean) value;
        }
        if (value instanceof String) {
            if ("true".equalsIgnoreCase((String) value)) {
                result = true;
            }
            if ("false".equalsIgnoreCase((String) value)) {
                result = false;
            }
        }
        return result;
    }

    /**
     * Get boolean value of key or def value.
     *
     * @param key key
     * @return boolean value of key
     */
    public boolean getBoolean(String key) {
        return getBoolean(key, false);
    }

    /**
     * Get int value of key.
     *
     * @param key      key
     * @param defValue defValue
     * @return Value of key
     */
    public int getInt(String key, int defValue) {
        Object value = this.get(key);
        if (value instanceof Integer) {
            return (int) value;
        }
        if (value instanceof String) {
            try {
                return Integer.parseInt((String) value);
            } catch (NumberFormatException exception) {
                LOGGER.debug(exception.getMessage(), exception);
            }
        }
        return defValue;
    }

    /**
     * Get int value of key or def value.
     *
     * @param key key
     * @return Value of key
     */
    public int getInt(String key) {
        return getInt(key, 0);
    }
}

